### PR TITLE
Feat: Only ask new questions on upgrade of apps

### PR DESCRIPTION
**Pull request by pmuir**
#### Submitter checklist

- [X] Change is code complete and matches issue description.
- [X] Change is covered by existing or new tests.

#### Description
Also adds support for batch mode

#### Which issue this PR fixes

fixes #

### CODE REVIEW

File: pkg/environments/gitops.go [hunk 1]
```diff
@@ -23,10 +23,22 @@ import (
 	helmchart "k8s.io/helm/pkg/proto/hapi/chart"
 )
 
+// PullRequestDetails is the details for creating a pull request
```
Comment (hferentschik) [inline pkg/environments/gitops.go]: 👍

File: pkg/apps/gitops.go [hunk 2]
```diff
@@ -21,9 +21,11 @@ type GitOpsOptions struct {
 
 // AddApp adds the app with version rooted in dir from the repository. An alias can be specified.
 func (o *GitOpsOptions) AddApp(app string, dir string, version string, repository string, alias string) error {
-	branchNameText := "add-app-" + app + "-" + version
-	title := fmt.Sprintf("Add %s %s", app, version)
-	message := fmt.Sprintf("Add app %s %s", app, version)
+	details := environments.PullRequestDetails{
```
Comment (hferentschik) [inline pkg/apps/gitops.go]: I like wrapping these up in `PullRequestDetails`

File: pkg/environments/gitops.go [hunk 3]
```diff
@@ -51,87 +63,20 @@ type EnvironmentPullRequestOptions struct {
...
-	gitInfo, err := gits.ParseGitURL(gitURL)
-	if err != nil {
-		return answer, err
-	}
-
-	if err != nil {
-		return answer, err
-	}
-	dir := filepath.Join(environmentsDir, gitInfo.Organisation, gitInfo.Name)
+func (o *EnvironmentPullRequestOptions) Create(env *jenkinsv1.Environment, environmentsDir string,
```
Comment (hferentschik) [inline pkg/environments/gitops.go]: Is is not really apps related right? It is a fallout of the additional structs you introduce, right?

File: pkg/jx/cmd/add_app.go [hunk 4]
```diff
@@ -74,15 +74,15 @@ func NewCmdAddApp(f Factory, in terminal.FileReader, out terminal.FileWriter, er
 		},
 	}
 
-	options.addFlags(cmd, kube.DefaultNamespace, "", "")
+	options.addFlags(cmd, kube.DefaultNamespace)
```
Comment (hferentschik) [inline pkg/jx/cmd/add_app.go]: > package cmd

Didn't you say that you started to have split up the _cmd_ package? Would it make sense to have this in a _add_ subdirectory? Or is the idea to restructure this in a big bang effort?

File: pkg/jx/cmd/upgrade_apps_test.go [hunk 5]
```diff
@@ -77,13 +80,372 @@ func TestUpgradeAppForGitOps(t *testing.T) {
...
+	assert.NoError(t, err)
+	assert.Equal(t, `name: testing
+species: human
+`, string(existingValues))
+}
+
+func TestUpgradeAppWithExistingAndDefaultAnswersForGitOps(t *testing.T) {
+	testOptions := cmd_test_helpers.CreateAppTestOptions(true, t)
+	defer func() {
+		err := testOptions.Cleanup()
```
Comment (hferentschik) [inline pkg/jx/cmd/upgrade_apps_test.go]: will that tear down any created directories, artefacts, etc, right?

Comment (hferentschik) [reply pkg/jx/cmd/upgrade_apps_test.go]: Seems quite a bit of work to set up a test. What's your experience with maintaining of these tests?

**Comment by hferentschik** [comment]
@pmuir looks like some sort of compilation problem atm

File: pkg/jx/cmd/add_app.go (hunk 4 above)
Comment (pmuir) [reply pkg/jx/cmd/add_app.go]: Yeah, we could do. I'll do that as a separate PR though!

File: pkg/jx/cmd/upgrade_apps_test.go (hunk 5 above)
Comment (pmuir) [reply pkg/jx/cmd/upgrade_apps_test.go]: > will that tear down any created directories, artefacts, etc, right?

Yes

Comment (pmuir) [reply pkg/jx/cmd/upgrade_apps_test.go]: > Seems quite a bit of work to set up a test. What's your experience with maintaining of these tests?

The thing that is very hard to maintain is the expect based tests - they just randomly go wrong sometimes :-( The rest is basically ok - I've tried to put what I can in helper methods that generically does stuff - then the test can focus on logic - the one you comment on the console stuff should go into the helper really, but the rest is really actually test specific - adding an app to do an upgrade against, running the upgrade, mocking the chart repo (via the `StubFetchChart`), running the console, validating the PR was created correctly.

File: pkg/environments/gitops.go (hunk 3 above)
Comment (pmuir) [reply pkg/environments/gitops.go]: It's part of a more generic piece of functionality that allows us to create PRs against GitOps environments. The `dev` environment is where we store info on installed apps when we're using gitops mode.

**Comment by pmuir** [comment]
/test
/hold
Holding PRs as we have a big refactor to get in first

**Comment by jstrachan** [comment]
@pmuir ah sorry I think I caused a merge confict

**Comment by pmuir** [comment]
No probs, will fix

File: pkg/environments/gitops.go (hunk 3 above)
Comment (hferentschik) [reply pkg/environments/gitops.go]: I see

**Pull request merged**